### PR TITLE
AMDGPU: Fix buffer load/store of pointers

### PR DESCRIPTION
**Pull request by arsenm**
Make sure we test all the address spaces since this support isn't
free in gisel.

### CODE REVIEW

**Comment by arsenm** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95379?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#95379** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95379?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#95378** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95378?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95377** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95377?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95376** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95376?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/CodeGen/AMDGPU/llvm.amdgcn.raw.ptr.buffer.store.ll [hunk 1]
```diff
@@ -424,6 +424,150 @@ define void @buffer_store_v2i64__voffset_add(ptr addrspace(8) inreg %rsrc, <2 x
   ret void
 }
 
+define void @buffer_store_p0__voffset_add(ptr addrspace(8) inreg %rsrc, ptr %data, i32 %voffset) #0 {
+  %voffset.add = add i32 %voffset, 60
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/llvm.amdgcn.raw.ptr.buffer.store.ll:428]: Missing checks?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/llvm.amdgcn.raw.ptr.buffer.store.ll:428]: For some reason this test isn't already autogenerated and I reverted too much of the generated checks

**Comment by krzysz00** [comment]
So, general question on this patch series:

Wouldn't it be more reasonable to, instead of having separate handling for all the possible register types, always do loads as `i8`, `i16`, `i32` `<2 x i32>`, `<3 x i32>, or `<4 x i32>` and then `bitcast`/`merge_values`/... the results back to their type?

Or at least to have that fallback path - if we don't know what a type is, load/store it as its bits?

(Then we wouldn't need to, for example, go back and add a `<16 x i8>` case if someone realizes they want that)

**Comment by arsenm** [comment]
That's what we've traditionally done and I think we should stop. We currently skip inserting the casts if the type is legal. 

It introduces extra bitcasts, which have a cost and increase pattern match complexity. We have a bunch of patterns that don't bother to look through the casts for a load/store

**Comment by krzysz00** [comment]
Yeah, makes sense.

... what prevents a match-bitwidth operator from existing?

Context from where I'm standing is that you should be able to `raw.buffer.load/store` any (non-aggregate, let's say, since that could be better handled in `addrspace(7)` handling) type you could `load` or `store`.

That is, `raw.ptr.buffer.load.i15` should work (as an i16 load that truncates) as should `raw.ptr.buffer.store.v8f32` (or `raw.ptr.buffer.store.i256`). Sure, the latter are two instructions long, but regular loads can regularize to multiple instructions just fine. 

My thoughts on how to implement that second behavior were to split the type into legal chunks and add in the offsets, and then merge/bitcast the values back.

**Comment by arsenm** [comment]
I don't think we should be trying to handle the unreasonable illegal types in the intrinsics themselves. Theoretically the intrinsic should correspond to direct support.

We would handle the ugly types in the fat pointer lowering in terms of the intrinsics.

**Comment by krzysz00** [comment]
On the other hand, it's a lot easier to handle ugly types down in instruction selection, where you get to play much more fast and loose with types.

And there are buffer uses that don't fit into the fat pointer use use case where we'd still want them to work.
For example, both `str
uct.ptr.bufferload.v6f16` and `struct.ptr.buffer.load.v3f32` should be a `buffer_load_dwordx3`, but I'm pretty sure 6 x half isn't a register type.

The load and store intrinsics are already overloaded to handle various {8, 16, ..., 128}-bit types, and it seems much cleaner to let it support any type of those lengths. It's just a load/store with somewhat weird indexing semantics, is all.

And then, since we're there ... `load i256, ptr addrspace(1) %p` legalizes to multiple instructions, and `{raw,struct}.ptr.buffer.load(ptr addrspace(8) %p, i32 %offset, ...)` should too. It's just a load, after all.

**Comment by arsenm** [comment]
> On the other hand, it's a lot easier to handle ugly types down in instruction selection, where you get to play much more fast and loose with types.

I think it's mostly easier to do this in the IR 

> 
> And there are buffer uses that don't fit into the fat pointer use use case where we'd still want them to work. For example, both `str uct.ptr.bufferload.v6f16` and `struct.ptr.buffer.load.v3f32` should be a `buffer_load_dwordx3`, but I'm pretty sure 6 x half isn't a register type.

Yes, we should just fix this one

> 
> The load and store intrinsics are already overloaded to handle various {8, 16, ..., 128}-bit types, and it seems much cleaner to let it support any type of those lengths. It's just a load/store with somewhat weird indexing semantics, is all.

Splitting is pretty ugly too, especially for a truly arbitrary type in legalization.

**Review by krzysz00 — APPROVED** [review]

**Pull request merged**